utilities/src/signals.rs

```diff
@@ -0,0 +1,140 @@
+///
+/// This is useful when no signal or complex shutdown handling is necessary (e.g., in a test environment).
+pub struct SimpleStoppable {
+    state: AtomicBool,
+}
+
+impl SimpleStoppable {
+    pub fn new() -> Arc<Self> {
+        Arc::new(Self { state: AtomicBool::new(false) })
+    }
+}
+
+impl Stoppable for SimpleStoppable {
+    fn stop(&self) {
+        self.state.store(true, Ordering::SeqCst);
+    }
+
+    fn is_stopped(&self) -> bool {
+        self.state.load(Ordering::SeqCst)
+    }
+}
+
+/// Helper for signal handling that implements the `Stoppable` trait.
+///
+/// This struct will set itself to "stopped" as soon as the process receives Ctrl+C.
+/// It can also be manually stopped (e.g., when the node encounters a fatal error).
+pub struct SignalHandler {
+    stopped: AtomicBool,
+    notify: Notify,
+}
+
+impl SignalHandler {
+    /// Spawns a background tasks that listens for Ctrl+C and returns `Self`.
+    pub fn new() -> Arc<Self> {
+        let obj = Arc::new(Self { stopped: AtomicBool::new(false), notify: Default::default() });
+
+        {
+            let obj = obj.clone();
+            tokio::spawn(async move {
+                obj.handle_signals().await;
+            });
+        }
+
+        obj
+    }
+
+    /// Logic for the background task that waits for a signal.
+    async fn handle_signals(&self) {
+        #[cfg(target_family = "unix")]
+        let signal_listener = async move {
+            use tokio::signal::unix::{SignalKind, signal};
+
+            // Handle SIGINT, SIGTERM, SIGQUIT, and SIGHUP.
+            let mut s_int = signal(SignalKind::interrupt())?;
+            let mut s_term = signal(SignalKind::terminate())?;
+            let mut s_quit = signal(SignalKind::quit())?;
+            let mut s_hup = signal(SignalKind::hangup())?;
+
+            tokio::select!(
+                _ = s_int.recv() => debug!("Received SIGINT"),
+                _ = s_term.recv() => debug!("Received SIGTERM"),
+                _ = s_quit.recv() => debug!("Received SIGQUIT"),
+                _ = s_hup.recv() => debug!("Received SIGHUP"),
+            );
+
+            std::io::Result::<()>::Ok(())
+        };
+
+        #[cfg(not(target_family = "unix"))]
+        let signal_listener = async move {
+            tokio::signal::ctrl_c().await?;
+            debug!("Received signal");
+
+            std::io::Result::<()>::Ok(())
+        };
+
+        // Block until the signal.
+        match signal_listener.await {
+            Ok(()) => {}
+            Err(error) => {
+                error!("tokio::signal encountered an error: {error}");
+            }
+        }
+
+        self.stop();
+    }
+
+    /// Blocks until the signal handler was invoked or the stopped flag was set some other way.
+    /// Note: This can only be called once, and must not be called concurrently.
+    pub async fn wait_for_signals(&self) {
+        while !self.is_stopped() {
+            self.notify.notified().await
+        }
```

Comment:
this seems a little bit convoluted; wouldn't `self.notify.notified().await` suffice?